File: src/utils/defenseTrainingGuidance.ts
import {
  buildPlayDescentLayout,
  countClearedStageNodes,
  findFrontierNodeId,
  isPlayDescentNodeUnlocked,
} from '@/components/play/defenseDescent/playDescentLayout';
import { isPlayMapProgressionGate } from '@/components/play/defenseDescent/playMapProgression';
import type { PlayMapBlock, PlayMapNode, PlayMapTier } from '@/platform/supabasePlayMap';

export type DefenseGuidanceReason = 'tutorial' | 'nextStage';

export type DefenseTrainingGuidance =
  | {
    kind: 'openDefense';
    tier: PlayMapTier;
    nodeId: string;
    nodeTitle: string;
    reason: DefenseGuidanceReason;
  }
  | { kind: 'openTraining' }
  | { kind: 'none' };

export interface ResolveDefenseTrainingGuidanceInput {
  isPremiumMember: boolean;
  blocks: readonly PlayMapBlock[];
  nodes: readonly PlayMapNode[];
  clearedNodeIds: ReadonlySet<string>;
  isEnglishCopy?: boolean;
}

const sortedTierBlocks = (
  blocks: readonly PlayMapBlock[],
  tier: PlayMapTier,
): PlayMapBlock[] =>
  blocks.filter((block) => block.tier === tier).sort((a, b) => a.sortOrder - b.sortOrder);

const findBasicTutorialNode = (
  blocks: readonly PlayMapBlock[],
  nodes: readonly PlayMapNode[],
): PlayMapNode | null => {
  const block0 = sortedTierBlocks(blocks, 'basic')[0];
  if (!block0) {
    return null;
  }
  return nodes.find((node) => node.blockId === block0.id && node.nodeKind === 'tutorial') ?? null;
};

const nodeDisplayTitle = (node: PlayMapNode, isEnglishCopy: boolean): string =>
  (isEnglishCopy ? node.titleEn : node.title);

const findNextUnclearedStageInBlock = (
  block: PlayMapBlock,
  nodes: readonly PlayMapNode[],
  clearedNodeIds: ReadonlySet<string>,
): PlayMapNode | null => {
  const stages = nodes
    .filter((node) => node.blockId === block.id && node.nodeKind === 'stage')
    .sort((a, b) => a.sortOrder - b.sortOrder);
  for (const stage of stages) {
    if (!clearedNodeIds.has(stage.id)) {
      return stage;
    }
  }
  return null;
};

const countTierClearedStages = (
  blocks: readonly PlayMapBlock[],
  nodes: readonly PlayMapNode[],
  tier: PlayMapTier,
  clearedNodeIds: ReadonlySet<string>,
): number => {
  const layout = buildPlayDescentLayout(blocks, nodes, tier);
  return countClearedStageNodes(layout, clearedNodeIds);
};

const resolvePlayableFrontier = (
  tier: PlayMapTier,
  blocks: readonly PlayMapBlock[],
  nodes: readonly PlayMapNode[],
  clearedNodeIds: ReadonlySet<string>,
  isPremiumMember: boolean,
  isEnglishCopy: boolean,
): Extract<DefenseTrainingGuidance, { kind: 'openDefense' }> | null => {
  const layout = buildPlayDescentLayout(blocks, nodes, tier, isEnglishCopy);
  const frontierId = findFrontierNodeId(layout.blocks, clearedNodeIds, isPremiumMember);
  if (!frontierId) {
    return null;
  }

  const frontierNode = nodes.find((node) => node.id === frontierId);
  if (!frontierNode) {
    return null;
  }

  if (
    isPlayMapProgressionGate(frontierNode)
    && !clearedNodeIds.has(frontierId)
    && isPlayDescentNodeUnlocked(frontierId, layout.blocks, clearedNodeIds, isPremiumMember)
  ) {
    return {
      kind: 'openDefense',
      tier,
      nodeId: frontierId,
      nodeTitle: nodeDisplayTitle(frontierNode, isEnglishCopy),
      reason: 'nextStage',
    };
  }

  return null;
};

const resolveTutorialGuidance = (
  tutorial: PlayMapNode,
  isEnglishCopy: boolean,
): Extract<DefenseTrainingGuidance, { kind: 'openDefense' }> => ({
  kind: 'openDefense',
  tier: 'basic',
  nodeId: tutorial.id,
  nodeTitle: nodeDisplayTitle(tutorial, isEnglishCopy),
  reason: 'tutorial',
});

export function resolveDefenseTrainingGuidance(
  input: ResolveDefenseTrainingGuidanceInput,
): DefenseTrainingGuidance {
  const {
    isPremiumMember,
    blocks,
    nodes,
    clearedNodeIds,
    isEnglishCopy = false,
  } = input;

  if (blocks.length === 0 || nodes.length === 0) {
    return { kind: 'none' };
  }

  const tutorial = findBasicTutorialNode(blocks, nodes);
  const tutorialCleared = tutorial != null && clearedNodeIds.has(tutorial.id);

  if (!isPremiumMember) {
    if (tutorial && !tutorialCleared) {
      return resolveTutorialGuidance(tutorial, isEnglishCopy);
    }

    const basicBlock0 = sortedTierBlocks(blocks, 'basic')[0];
    if (basicBlock0) {
      const nextStage = findNextUnclearedStageInBlock(basicBlock0, nodes, clearedNodeIds);
      if (nextStage) {
        return {
          kind: 'openDefense',
          tier: 'basic',
          nodeId: nextStage.id,
          nodeTitle: nodeDisplayTitle(nextStage, isEnglishCopy),
          reason: 'nextStage',
        };
      }
    }

    return { kind: 'openTraining' };
  }

  const basicCleared = countTierClearedStages(blocks, nodes, 'basic', clearedNodeIds);
  const advancedCleared = countTierClearedStages(blocks, nodes, 'advanced', clearedNodeIds);
  const preferredTier: PlayMapTier = advancedCleared > basicCleared ? 'advanced' : 'basic';
  const tiersToTry: PlayMapTier[] = preferredTier === 'basic'
    ? ['basic', 'advanced']
    : ['advanced', 'basic'];

  for (const tier of tiersToTry) {
    if (tier === 'basic' && tutorial && !tutorialCleared) {
      return resolveTutorialGuidance(tutorial, isEnglishCopy);
    }

    const frontierGuidance = resolvePlayableFrontier(
      tier,
      blocks,
      nodes,
      clearedNodeIds,
      true,
      isEnglishCopy,
    );
    if (frontierGuidance) {
      return frontierGuidance;
    }
  }

  return { kind: 'openTraining' };
}

export function buildDefenseNodeHash(nodeId: string): string {
  return `#phrase-defense?nodeId=${encodeURIComponent(nodeId)}`;
}

export const TRAINING_ROUTE_HASH = '#training';

export function defenseGuidancePrimaryLabel(
  guidance: Extract<DefenseTrainingGuidance, { kind: 'openDefense' }>,
  isEnglishCopy: boolean,
): string {
  if (guidance.reason === 'tutorial') {
    return isEnglishCopy ? 'Start first-time setup' : 'はじめての設定を始める';
  }
  return isEnglishCopy ? 'Continue Phrase Defense' : 'フレーズディフェンスを続ける';
}

export function defenseGuidanceBodyCopy(
  guidance: DefenseTrainingGuidance,
  isEnglishCopy: boolean,
): string | null {
  if (guidance.kind === 'openDefense') {
    const quotedTitle = isEnglishCopy
      ? `"${guidance.nodeTitle}"`
      : `「${guidance.nodeTitle}」`;
    if (guidance.reason === 'tutorial') {
      return isEnglishCopy
        ? `Set up your input with ${quotedTitle}.`
        : `${quotedTitle}で入力設定を行いましょう。`;
    }
    return isEnglishCopy
      ? `Next up: ${quotedTitle}`
      : `次は${quotedTitle}です。`;
  }
  if (guidance.kind === 'openTraining') {
    return isEnglishCopy
      ? 'Keep building skills in Training.'
      : 'トレーニングでスキルを伸ばしましょう。';
  }
  return null;
}

File: ios/Jazzify/Main/DefenseTrainingGuidance.swift
import Foundation

enum DefenseGuidanceReason: Equatable, Sendable {
    case tutorial
    case nextStage
}

enum DefenseTrainingGuidance: Equatable, Sendable {
    case openDefense(
        tier: PlayMapTier,
        nodeId: UUID,
        nodeTitle: String,
        reason: DefenseGuidanceReason
    )
    case defenseBlockComplete
    case openTraining
    case none
}

enum DefenseTrainingPromptKind: Equatable, Sendable {
    case resume
    case nextStep
}

enum DefenseTrainingGuidanceResolver {
    private static func sortedTierBlocks(
        _ blocks: [PlayMapBlock],
        tier: PlayMapTier
    ) -> [PlayMapBlock] {
        blocks
            .filter { $0.tier == tier }
            .sorted { $0.sortOrder < $1.sortOrder }
    }

    private static func findBasicTutorialNode(
        blocks: [PlayMapBlock],
        nodes: [PlayMapNode]
    ) -> PlayMapNode? {
        guard let block0 = sortedTierBlocks(blocks, tier: .basic).first else {
            return nil
        }
        return nodes.first { $0.blockId == block0.id && $0.nodeKind == .tutorial }
    }

    private static func nodeDisplayTitle(_ node: PlayMapNode, locale: AppLocale) -> String {
        node.localizedTitle(locale)
    }

    private static func isBlockFullyCleared(
        block: PlayMapBlock,
        nodes: [PlayMapNode],
        clearedNodeIds: Set<UUID>
    ) -> Bool {
        let stages = nodes.filter { $0.blockId == block.id && $0.nodeKind == .stage }
        guard !stages.isEmpty else { return false }
        return stages.allSatisfy { clearedNodeIds.contains($0.id) }
    }

    private static func findNextUnclearedStageInBlock(
        block: PlayMapBlock,
        nodes: [PlayMapNode],
        clearedNodeIds: Set<UUID>
    ) -> PlayMapNode? {
        let stages = nodes
            .filter { $0.blockId == block.id && $0.nodeKind == .stage }
            .sorted { $0.sortOrder < $1.sortOrder }
        return stages.first { !clearedNodeIds.contains($0.id) }
    }

    private static func countTierClearedStages(
        blocks: [PlayMapBlock],
        nodes: [PlayMapNode],
        tier: PlayMapTier,
        clearedNodeIds: Set<UUID>
    ) -> Int {
        let layout = DefenseDescentLayoutBuilder.build(
            blocks: blocks,
            nodes: nodes,
            tier: tier
        )
        return DefenseDescentAccess.countClearedStageNodes(in: layout, clearedNodeIds: clearedNodeIds)
    }

    private static func resolvePlayableFrontier(
        tier: PlayMapTier,
        blocks: [PlayMapBlock],
        nodes: [PlayMapNode],
        clearedNodeIds: Set<UUID>,
        isPremium: Bool,
        locale: AppLocale
    ) -> DefenseTrainingGuidance? {
        let layout = DefenseDescentLayoutBuilder.build(
            blocks: blocks,
            nodes: nodes,
            tier: tier
        )
        guard let frontierId = DefenseDescentAccess.findFrontierNodeId(
            blockLayouts: layout.blocks,
            clearedNodeIds: clearedNodeIds,
            isPremium: isPremium
        ) else {
            return nil
        }
        guard let frontierNode = nodes.first(where: { $0.id == frontierId }) else {
            return nil
        }
        guard PlayMapProgression.isProgressionGate(frontierNode),
              !clearedNodeIds.contains(frontierId),
              DefenseDescentAccess.isNodeUnlocked(
                nodeId: frontierId,
                blockLayouts: layout.blocks,
                clearedNodeIds: clearedNodeIds,
                isPremium: isPremium
              ) else {
            return nil
        }
        return .openDefense(
            tier: tier,
            nodeId: frontierId,
            nodeTitle: nodeDisplayTitle(frontierNode, locale: locale),
            reason: .nextStage
        )
    }

    private static func resolveTutorialGuidance(
        tutorial: PlayMapNode,
        locale: AppLocale
    ) -> DefenseTrainingGuidance {
        .openDefense(
            tier: .basic,
            nodeId: tutorial.id,
            nodeTitle: nodeDisplayTitle(tutorial, locale: locale),
            reason: .tutorial
        )
    }

    static func resolve(
        isPremium: Bool,
        blocks: [PlayMapBlock],
        nodes: [PlayMapNode],
        clearedNodeIds: Set<UUID>,
        locale: AppLocale
    ) -> DefenseTrainingGuidance {
        guard !blocks.isEmpty, !nodes.isEmpty else {
            return .none
        }

        let tutorial = findBasicTutorialNode(blocks: blocks, nodes: nodes)
        let tutorialCleared = tutorial.map { clearedNodeIds.contains($0.id) } ?? true

        if !isPremium {
            if let tutorial, !tutorialCleared {
                return resolveTutorialGuidance(tutorial: tutorial, locale: locale)
            }
            let basicBlock0 = sortedTierBlocks(blocks, tier: .basic).first
            if let basicBlock0,
               let nextStage = findNextUnclearedStageInBlock(
                block: basicBlock0,
                nodes: nodes,
                clearedNodeIds: clearedNodeIds
               ) {
                return .openDefense(
                    tier: .basic,
                    nodeId: nextStage.id,
                    nodeTitle: nodeDisplayTitle(nextStage, locale: locale),
                    reason: .nextStage
                )
            }
            if let basicBlock0,
               isBlockFullyCleared(block: basicBlock0, nodes: nodes, clearedNodeIds: clearedNodeIds) {
                return .defenseBlockComplete
            }
            return .openTraining
        }

        let basicCleared = countTierClearedStages(
            blocks: blocks,
            nodes: nodes,
            tier: .basic,
            clearedNodeIds: clearedNodeIds
        )
        let advancedCleared = countTierClearedStages(
            blocks: blocks,
            nodes: nodes,
            tier: .advanced,
            clearedNodeIds: clearedNodeIds
        )
        let preferredTier: PlayMapTier = advancedCleared > basicCleared ? .advanced : .basic
        let tiersToTry: [PlayMapTier] = preferredTier == .basic ? [.basic, .advanced] : [.advanced, .basic]

        for tier in tiersToTry {
            if tier == .basic, let tutorial, !tutorialCleared {
                return resolveTutorialGuidance(tutorial: tutorial, locale: locale)
            }
            if let frontierGuidance = resolvePlayableFrontier(
                tier: tier,
                blocks: blocks,
                nodes: nodes,
                clearedNodeIds: clearedNodeIds,
                isPremium: true,
                locale: locale
            ) {
                return frontierGuidance
            }
        }

        return .openTraining
    }

    static func loadTodayStreakUpdated(profile: Profile?) async -> Bool {
        let timezone = TrainingActivity.resolveUserTimezone(profile: profile)
        let todayKey = TrainingActivity.localDateKey(Date(), timezone: timezone)
        let days = (try? await SupabaseService.shared.fetchTrainingActivityDays(timezone: timezone)) ?? []
        return TrainingActivity.isTodayStreakUpdated(activeDays: Set(days), todayKey: todayKey)
    }

    static func sheetTitle(
        for guidance: DefenseTrainingGuidance,
        kind: DefenseTrainingPromptKind,
        locale: AppLocale,
        todayStreakUpdated: Bool = false
    ) -> String {
        switch kind {
        case .resume:
            switch guidance {
            case .openTraining:
                if todayStreakUpdated {
                    return locale == .ja ? "トレーニングを続けますか？" : "Keep going in Training?"
                }
                return locale == .ja ? "今日の連続記録を更新しますか？" : "Update today's streak?"
            default:
                return locale == .ja ? "続きから再開しますか？" : "Continue where you left off?"
            }
        case .nextStep:
            switch guidance {
            case .openTraining:
                return locale == .ja ? "お疲れさまでした！" : "Nice work!"
            case .defenseBlockComplete:
                return locale == .ja ? "第1階層をクリアしました！" : "You cleared Floor 1!"
            default:
                return locale == .ja ? "次に進みますか？" : "Ready for the next step?"
            }
        }
    }

    static func blockCompleteTrialLabel(locale: AppLocale) -> String {
        locale == .ja ? "7日無料で続きを試す" : "Try the next tier free for 7 days"
    }

    static func blockCompleteSoftLandingLabel(locale: AppLocale) -> String {
        locale == .ja ? "コードランを無料で始める" : "Start Chord Run free"
    }

    static func blockCompleteBodyCopy(locale: AppLocale) -> String {
        locale == .ja
            ? "第1階層をクリアしました。Advanced と全ステージはプレミアムで解放できます。"
            : "You cleared Floor 1. Unlock Advanced and all stages with Premium."
    }

    static func blockCompleteNextStepLabel(locale: AppLocale) -> String {
        locale == .ja ? "次のステップを見る" : "See what's next"
    }

    static func primaryLabel(
        for guidance: DefenseTrainingGuidance,
        locale: AppLocale,
        todayStreakUpdated: Bool = false
    ) -> String? {
        switch guidance {
        case .openDefense(_, _, _, let reason):
            switch reason {
            case .tutorial:
                return locale == .ja ? "はじめての設定を始める" : "Start first-time setup"
            case .nextStage:
                return locale == .ja ? "フレーズディフェンスを続ける" : "Continue Phrase Defense"
            }
        case .defenseBlockComplete:
            return blockCompleteNextStepLabel(locale: locale)
        case .openTraining:
            if todayStreakUpdated {
                return locale == .ja ? "トレーニングへ" : "Go to Training"
            }
            return locale == .ja ? "今日の連続記録を更新" : "Update today's streak"
        case .none:
            return nil
        }
    }

    static func bodyCopy(
        for guidance: DefenseTrainingGuidance,
        locale: AppLocale,
        todayStreakUpdated: Bool = false
    ) -> String? {
        switch guidance {
        case .openDefense(_, _, let nodeTitle, let reason):
            let quoted = locale == .ja ? "「\(nodeTitle)」" : "\"\(nodeTitle)\""
            switch reason {
            case .tutorial:
                return locale == .ja
                    ? "\(quoted)で入力設定を行いましょう。"
                    : "Set up your input with \(quoted)."
            case .nextStage:
                return locale == .ja ? "次は\(quoted)です。" : "Next up: \(quoted)"
            }
        case .defenseBlockComplete:
            return blockCompleteBodyCopy(locale: locale)
        case .openTraining:
            if todayStreakUpdated {
                return locale == .ja
                    ? "今日の連続記録は更新済みです。さらにトレーニングでスキルを伸ばしましょう。"
                    : "Today's streak is already updated. Keep building skills in Training."
            }
            return locale == .ja
                ? "今日の連続記録はまだ更新されていません。トレーニングで更新しましょう。"
                : "Today's training streak is not updated yet. Play Training to keep it going."
        case .none:
            return nil
        }
    }
}
